core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSail.java

```diff
@@ -307,11 +322,22 @@ public void shutDown() throws SailException {
 			shapesRepo = null;
 		}
 
+		if (executorService != null) {
+			executorService.shutdown();
+		}
+
 		initialized.set(false);
 		nodeShapes = Collections.emptyList();
 		super.shutDown();
 	}
 
+	public synchronized <T> Future<T> submitRunnableToExecutorService(Callable<T> runnable) {
+		if (executorService == null) {
+			executorService = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() * 2);
```

Comment:
Lazy init of the executor service, so it's only created when we first need parallel validation.